journal/202308.md

```diff
@@ -0,0 +1,870 @@
+
+
+### dcrwallet
+
+_[dcrwallet](https://github.com/decred/dcrwallet) is a wallet server used by command-line and graphical wallet apps._
+
+June-July changes [backported](https://github.com/decred/dcrwallet/pull/2277) for the next v1.8.x release:
+
+- [Warn users](https://github.com/decred/dcrwallet/commit/1c1bcfbd5aa7ca0b1cbad06b4f1bb6b8c2af5aa5) about VSP fee payment errors caused by the wallet being locked. This will allow users to react if a locked wallet blocks VSP fee payments. Upon seeing the error, the user should unlock the wallet and try again to prevent missed tickets.
+- Fixed broken [retrying logic](https://github.com/decred/dcrwallet/commit/07854e3746aeef2de88488ae5cc87ee6a092387b) for errored VSP ticket purchases. This issue affected many users. Any ticket purchases which worked first time were unaffected, but any purchases which ran into an error were not being retried, which would lead to tickets not being added to the VSP and votes being missed. Thankfully, re-registering the ticket with a different VSP was a workaround, but it was poor UX.
+- Fixed a bug where the [fee payment status check](https://github.com/decred/dcrwallet/commit/c4c6c4213da7cde5ecdc97fb5347a1c7bc3aeae2) could be unreasonably delayed. The VSP client delays actions by some random duration in order to help protect privacy. With this incorrect delay calculation the wallet would report fee payments as being unconfirmed for up to 2 hours until the next check happens, when actually the fee has been sent and confirmed. Now the user wallet will update sooner.
+- Require VSP clients to [wait for the VSP to confirm fee payments](https://github.com/decred/dcrwallet/commit/7c551ce6b95456b6ec1a174b5529641dd133230a) before considering a VSP ticket fully purchased. This eliminates multiple problematic edge cases caused by users broadcasting their own fee transactions, such as paying too low of a fee or sending payment to the wrong address. It is a preventative fix for extremely unlikely cases which have not been seen in the wild.
+
+New changes merged in `master` and backported for the next v1.8.x release:
+
+- Fixed a bug where ticket purchase transactions could be sent to the VSP client before they were mined, resulting in the VSP not [calculating the block heights](https://github.com/decred/dcrwallet/pull/2270) where the ticket would be live/expired. Without knowing if the ticket was live or expired, the VSP wouldn't vote on the ticket, and would result in a missed ticket. Backported change is [here](https://github.com/decred/dcrwallet/commit/1319da0b05d4a9715b045e847ff58b442f407f39).
+  - { _@jholdstock, is this correct? What happened previously when `fp.ticketLive == 0` or `fp.ticketExpires == 0`? I can't figure out exactly what users would experience. Also, how often would unmined transactions reach the VSP? Seems like since the mempool travels faster than blocks, it would happen quite a lot, but VSPs don't miss many tickets._ }
```

Comment:
@jholdstock please check this and the vspd section below if you have time.

Reply:
> how often would unmined transactions reach the VSP? Seems like since the mempool travels faster than blocks, it would happen quite a lot

Wallets send their ticket/fee transactions directly to the VSP, so mempool/mining/block propagation isn't actually relevant for this one.

In code an unmined transaction is typically given a height of 0, its a totally normal thing to occur. The issue was that the 0 height was never being updated, even after the transaction was mined and the real height was known.

The impact of this bug was pretty much the same as the other one described above:

> fee payment status check could be unreasonably delayed ... With this incorrect delay calculation the wallet would report fee payments as being unconfirmed for up to 2 hours until the next check happens, when actually the fee has been sent and confirmed. Now the user wallet will update sooner.

There was no risk of tickets not being added to wallets or of votes being missed.

Reply:
Thanks. So we wrongly assumed the VSP would not vote/miss the ticket, that is now fixed. With your comment and a re-read of 2270 this is my understanding so far:

- The issue was on the client (end user) side, I didn't see related bugs/fixes in vspd. Therefore it was not causing VSP to not vote/miss tickets.
- It was not causing the end user (dcrwallet + VSP client) to "lose" the ticket tx or fee tx.
- It was not causing the VSP client to send incorrect data to VSP server.
- The only impact was an unreasonably long time to consider the ticket fee as paid (and user confusion resulting from that) - similar to what we said about 2269. The cause of that was live/expiry heights never being updated after the ticket was mined.

Correct so far?

What remains unclear to me is:

1. Did it happen always or in some cases?
2. Did it eventually "resolve itself" after some time, i.e. did the fee tx eventually show up as confirmed, or did the buggy status persist forever?

Reply:
Thanks! All clear now. Incorporated in 34f45a26